compiler/rustc_infer/src/infer/opaque_types.rs

```diff
@@ -381,8 +387,34 @@ impl<'tcx> InferCtxt<'tcx> {
             // Anonymous `impl Trait`
             hir::OpaqueTyOrigin::FnReturn(parent) => parent == parent_def_id,
             // Named `type Foo = impl Bar;`
-            hir::OpaqueTyOrigin::TyAlias => {
-                may_define_opaque_type(self.tcx, parent_def_id, opaque_hir_id)
+            hir::OpaqueTyOrigin::TyAlias { in_assoc_ty } => {
+                if in_assoc_ty {
+                    self.tcx.opaque_types_defined_by(parent_def_id).transpose_iter().any(|t| {
+                        trace!(a=?t.skip_binder().skip_binder().kind());
+                        trace!(b=?ty.kind());
+
+                        let res = self.commit_if_ok(|_| {
```

Comment:
Isn't this essentially equivalent to `can_eq`? Or are the side-effect actually visible from this commit? I'm not sure if I totally understand what's going on here.

Reply:
yea, this is equivalent to `can_eq`, except that it also does a check for no obligations being emitted.